R/data.table.R

```diff
@@ -1943,13 +1945,14 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
   }
   
   # convert columns to common class before handing over to C
-  col.classes = unlist(lapply(X, class))
+  col.classes = unlist(lapply(X[-rownames], class))
   class.order = c("logical"=1L, "integer"=2L, "numeric"=3L, "complex"=4L,
                   "character"=5L, "raw"=6L, "list"=7L)
   if (length(unique(col.classes)) > 1L) { 
     target.class = names(which.max(class.order[unique(col.classes)]))
     if (target.class == "raw") target.class = "character" # to match behaviour of as.matrix.data.frame
     for (col in which(col.classes != target.class)) {
+      ifelse(col >= rownames, col + 1L, col)
```

Comment:
This line is not doing anything, computation is lost.
If we still need it for something, please add tests so we can ensure it is being used.
Moreover I would personally avoid using `ifelse`, and here, in data.table internals where we operate on metadata I would even avoid using `fifelse`. Very good option is
```r
#ans = ifelse(col >= rownames, col + 1L, col)
ans = col
ans[col >= rownames] = ans+1L
```

Reply:
Good catch, this was causing an error when the rownames column was in the middle of the data.table and the columns needed to be typecast. Assuming this logic stays the same when I revisit all the column checks, I'll need to add a test to cover this case.